iceoryx_posh/source/runtime/posh_runtime_impl.cpp

```diff
@@ -282,6 +282,152 @@ PoshRuntimeImpl::requestSubscriberFromRoudi(const IpcMessage& sendBuffer) noexce
 }
 
+popo::ClientPortUser::MemberType_t* PoshRuntimeImpl::getMiddlewareClient(const capro::ServiceDescription& service,
+                                                                         const popo::ClientOptions& clientOptions,
+                                                                         const PortConfigInfo& portConfigInfo) noexcept
+{
+    IpcMessage sendBuffer;
+    sendBuffer << IpcMessageTypeToString(IpcMessageType::CREATE_CLIENT) << m_appName
+               << static_cast<cxx::Serialization>(service).toString() << clientOptions.serialize().toString()
+               << static_cast<cxx::Serialization>(portConfigInfo).toString();
+
+    auto maybeClient = requestClientFromRoudi(sendBuffer);
+    if (maybeClient.has_error())
+    {
+        switch (maybeClient.get_error())
+        {
+        case IpcMessageErrorType::CLIENT_LIST_FULL:
+            LogWarn() << "Could not create client with service description '" << service
+                      << "' as we are out of memory for clients.";
+            errorHandler(Error::kPOSH__RUNTIME_ROUDI_OUT_OF_CLIENTS, nullptr, iox::ErrorLevel::SEVERE);
+            break;
+        case IpcMessageErrorType::REQUEST_CLIENT_WRONG_IPC_MESSAGE_RESPONSE:
+            LogWarn() << "Could not create client with service description '" << service
+                      << "'; received wrong IPC channel response.";
+            errorHandler(Error::kPOSH__RUNTIME_ROUDI_REQUEST_CLIENT_WRONG_IPC_MESSAGE_RESPONSE,
+                         nullptr,
+                         iox::ErrorLevel::SEVERE);
+            break;
+        default:
+            LogWarn() << "Unknown error occurred while creating client with service description '" << service << "'";
+            errorHandler(Error::kPOSH__RUNTIME_CLIENT_PORT_CREATION_UNKNOWN_ERROR, nullptr, iox::ErrorLevel::SEVERE);
+            break;
+        }
+        return nullptr;
+    }
+    return maybeClient.value();
+}
+
+cxx::expected<popo::ClientPortUser::MemberType_t*, IpcMessageErrorType>
+PoshRuntimeImpl::requestClientFromRoudi(const IpcMessage& sendBuffer) noexcept
+{
+    IpcMessage receiveBuffer;
+    if (sendRequestToRouDi(sendBuffer, receiveBuffer) && (3U == receiveBuffer.getNumberOfElements()))
+    {
+        std::string IpcMessage = receiveBuffer.getElementAtIndex(0U);
+
+        if (stringToIpcMessageType(IpcMessage.c_str()) == IpcMessageType::CREATE_CLIENT_ACK)
+        {
+            rp::BaseRelativePointer::id_t segmentId{0U};
+            cxx::convert::fromString(receiveBuffer.getElementAtIndex(2U).c_str(), segmentId);
+            rp::BaseRelativePointer::offset_t offset{0U};
+            cxx::convert::fromString(receiveBuffer.getElementAtIndex(1U).c_str(), offset);
+            auto ptr = rp::BaseRelativePointer::getPtr(segmentId, offset);
+            return cxx::success<popo::ClientPortUser::MemberType_t*>(
+                reinterpret_cast<popo::ClientPortUser::MemberType_t*>(ptr));
+        }
+    }
+    else
```

Comment:
I think it may not be a technically error but it is somehow unclean.

Assume `sendRequestToRouDi` fails and returns false but the method changed already parts of the `receiveBuffer`. So we have somehow a `receiveBuffer` with crap content and go here into the else branch and check if this crap buffer has 2 elements. This could be successful but would somehow mean undefined behavior.

To fix this possible issue, could you please do an early return
````
if ( sendRequestToRouDi(sendBuffer, receiveBuffer) ) {
  // here we require an additional error code we can return when we are unable to send requests to roudi.
}
```

and rework the `if` `else` a bit so that you handle explicitly `numberOfElements` `2U` and `3U`.

Reply:
@elfenpiff done with this commit https://github.com/eclipse-iceoryx/iceoryx/pull/1137/commits/2a1280d3c7e586981ac8bc50b0d87c1c17c3d528 on #1137